### ISSUE TITLE
Game crashes when holding down keys during a loading screen

### ISSUE THREAD

**Issue by Gogsi** [crosire/scripthookvdotnet#1188]
Please note: **ScriptHookVDotNet is NOT ScriptHookV**!

Issue only occurs when ScriptHookVDotNet.dll is installed

-----

**Version**
Both latest release 3.6.0 and latest main commit f75c4a5

**Description**
This has been known for years in the speedrunning community but I haven't had the time to investigate it. Originally it was believed that holding down any character switch key during a loading screen causes ScriptHookVDotNet to crash the game (this is used in a glitch called [savewarping](https://www.youtube.com/watch?v=J5VYV50cJPU)).

I decided to investigate because I'm planning on using ScriptHookVDotNet for a project. It seems that the game will crash if you hold down any button. This is related to the console and I've pasted an Event Viewer error below (couldn't debug it in VS because the anticheat/DRM crashes the game if you have a debugger attached during a loading screen regardless).

```
Application: GTA5.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AppDomainUnloadedException
   at SHVDN.Console.DoKeyEvent(System.Windows.Forms.Keys, Boolean)
   at <Module>.?A0xddd40208.ScriptHookVDotnet_ManagedKeyboardMessage(UInt32, Boolean, Boolean, Boolean, Boolean)
```

It doesn't crash 100% of the time, I loaded successfully once with F6 and twice with U. I tried adding try/catches for the exception but it didn't seem to solve the issue. I'll leave this to someone more experienced to investigate I think.

**Comment by kagikn** [comment]
Man, I wish someone had been found a way to disable anti-debug features that can happen when the game is in the loading screen phase. In that case, the game utilizes them using `BeingDebugged` in the PEB struct (the game occasionally reads the flag when the game is opening the pause menu).

**Comment by ranstar74** [comment]
> Man, I wish someone had found ways to disable anti-debug features that can happen when the game is in the loading screen phase. In that case, the game utilizes them using `BeingDebugged` in the PEB struct (the game occasionally reads the flag when the game is opening the pause menu as well).

You may try scylla hide, I've been using it with C++ (without script hook, though). Game does crash eventually but most of the time it works fine, been debugging for hours without any issue.

**Comment by Gogsi** [comment]
I have no issues running the game through Project 1.27 with build f75c4a5. Tried it on two different devices and installed a random mod to see that it's not caused by loading mods. Perhaps you have a mod that isn't compatible with 1.27?

![image](https://user-images.githubusercontent.com/1671056/227866528-b630cef5-4b96-46e0-913f-1d120d8c771d.png)

Worth noting that the hold key crash also happens on the latest patch, so the issue is not dependent on game version

**Comment by Sardelka9515** [comment]
I believe this issue exisited long before, but now that the memory scanning is done when loading appdomain, the problem surfaces
The static gc reference `ScriptHookVDotNet::domain` and `ScriptHookVDotNet::console` wasn't set to null before unloading, which makes them points to remoting objects in unloaded AppDomain, causing `AppDomainUnloadedException` in the keyboard handler. It wan't that noticeable before becuase AppDomain unload/reload itself doesn't take much time.

Porbaly can be fixed by setting them to null before unloading 6b6191f4c2b8f98f8fa362a72f0285eb216089b2, or just catch the exception.

**Comment by Gogsi** [comment]
Ah, I wasn't catching for all occurrences of console and domain, that's why it wasn't working! Can confirm that both catching the exception and setting them to null work. The latter seems cleaner to me

**Comment by kagikn** [comment]
I haven't managed to run the game through Project 1.27 yet, but still I could reproduce the issue in the latest patch b2845 before merging #1192.

**Comment by kagikn** [comment]
> Just noting that the issue seems to be fixed with #1192 but out of nowhere I've started to experience the Project 1.27 crash on one of my two devices. Really weird because I'm almost confident it worked before when I tested [f75c4a5](https://github.com/crosire/scripthookvdotnet/commit/f75c4a5b12f23762ac543bccc0a6d51310dd22a4). On my other device both 3.6.0, [f75c4a5](https://github.com/crosire/scripthookvdotnet/commit/f75c4a5b12f23762ac543bccc0a6d51310dd22a4) and [5c368c3](https://github.com/crosire/scripthookvdotnet/commit/5c368c38e9236d33c6e01ed9f1286b02784bf993) work fine. I'll probably open another issue about it

I just Identified what commit made SHVDN being imcompatible with Project 1.27. It's ee58753. You can track the issue in #1202.
